File: src/accessors.zig
const std = @import("std");

// Env-variable access patterns, grouped by language and keyed by file
// extension. The scanner matches `prefix` at a position, then.patterns the key
// using .pattern`. This replaces convention-based suffix matching (`_ENV`) with
// intent-based matching: a token is an env var because it is read through a
// known environment accessor, not because of how it is spelled.
//
// WHAT THIS CANNOT CATCH (true of every language, lexical limit):
//   - destructuring:   const { DATABASE_URL } = process.env
//   - aliasing:        const e = process.env; e.FOO
//   - dynamic keys:    process.env[name]        os.getenv(var)

pub const Pattern = enum {
    ident, // bareword identifier follows the prefix:        process.env.FOO
    quoted, // first string literal ("..."/'...') follows:    getenv("FOO")
    braced, // token (bareword or quoted) up to '}':          $ENV{FOO}  ${FOO}
    parened, // bareword up to ')':                            $env(FOO)
};

pub const Accessor = struct {
    prefix: []const u8,
    pattern: Pattern,
};

const javascript = [_]Accessor{
    .{ .prefix = "process.env.", .pattern = .ident },
    .{ .prefix = "process.env[", .pattern = .quoted },
    .{ .prefix = "import.meta.env.", .pattern = .ident },
    .{ .prefix = "import.meta.env[", .pattern = .quoted },
};

const deno_bun = [_]Accessor{
    .{ .prefix = "Deno.env.get(", .pattern = .quoted },
    .{ .prefix = "Deno.env.has(", .pattern = .quoted },
    .{ .prefix = "Bun.env.", .pattern = .ident },
    .{ .prefix = "Bun.env[", .pattern = .quoted },
};

const python = [_]Accessor{
    .{ .prefix = "os.getenv(", .pattern = .quoted },
    .{ .prefix = "os.environ[", .pattern = .quoted },
    .{ .prefix = "os.environ.get(", .pattern = .quoted },
    .{ .prefix = "os.environ.setdefault(", .pattern = .quoted },
    .{ .prefix = "getenv(", .pattern = .quoted },
    .{ .prefix = "environ[", .pattern = .quoted },
    .{ .prefix = "environ.get(", .pattern = .quoted },
};

const go = [_]Accessor{
    .{ .prefix = "os.Getenv(", .pattern = .quoted },
    .{ .prefix = "os.LookupEnv(", .pattern = .quoted },
};

const rust = [_]Accessor{
    .{ .prefix = "std::env::var(", .pattern = .quoted },
    .{ .prefix = "std::env::var_os(", .pattern = .quoted },
    .{ .prefix = "env::var(", .pattern = .quoted },
    .{ .prefix = "env::var_os(", .pattern = .quoted },
    .{ .prefix = "env!(", .pattern = .quoted },
    .{ .prefix = "option_env!(", .pattern = .quoted },
};

const ruby = [_]Accessor{
    .{ .prefix = "ENV[", .pattern = .quoted },
    .{ .prefix = "ENV.fetch(", .pattern = .quoted },
};

const zig = [_]Accessor{
    .{ .prefix = "std.posix.getenv(", .pattern = .quoted },
    .{ .prefix = "posix.getenv(", .pattern = .quoted },
    .{ .prefix = "std.process.getEnvVarOwned(", .pattern = .quoted },
    .{ .prefix = "std.process.hasEnvVar(", .pattern = .quoted },
    .{ .prefix = "std.process.hasEnvVarConstant(", .pattern = .quoted },
    .{ .prefix = "std.process.parseEnvVarInt(", .pattern = .quoted },
};

const java = [_]Accessor{
    .{ .prefix = "System.getenv(", .pattern = .quoted },
};

const scala = [_]Accessor{
    .{ .prefix = "sys.env(", .pattern = .quoted },
    .{ .prefix = "sys.env.get(", .pattern = .quoted },
    .{ .prefix = "System.getenv(", .pattern = .quoted },
};

const clojure = [_]Accessor{
    .{ .prefix = "System/getenv ", .pattern = .quoted },
    .{ .prefix = "(getenv ", .pattern = .quoted },
};

const c = [_]Accessor{
    .{ .prefix = "getenv(", .pattern = .quoted },
    .{ .prefix = "secure_getenv(", .pattern = .quoted },
};

const cpp = [_]Accessor{
    .{ .prefix = "std::getenv(", .pattern = .quoted },
    .{ .prefix = "getenv(", .pattern = .quoted },
    .{ .prefix = "secure_getenv(", .pattern = .quoted },
};

const dotnet = [_]Accessor{
    .{ .prefix = "Environment.GetEnvironmentVariable(", .pattern = .quoted },
    .{ .prefix = "System.Environment.GetEnvironmentVariable(", .pattern = .quoted },
};

const visualbasic = [_]Accessor{
    .{ .prefix = "Environment.GetEnvironmentVariable(", .pattern = .quoted },
    .{ .prefix = "Environ(", .pattern = .quoted },
};

const php = [_]Accessor{
    .{ .prefix = "getenv(", .pattern = .quoted },
    .{ .prefix = "$_ENV[", .pattern = .quoted },
    .{ .prefix = "$_SERVER[", .pattern = .quoted },
};

const perl = [_]Accessor{
    .{ .prefix = "$ENV{", .pattern = .braced },
};

const powershell = [_]Accessor{
    .{ .prefix = "$env:", .pattern = .ident },
    .{ .prefix = "[Environment]::GetEnvironmentVariable(", .pattern = .quoted },
    .{ .prefix = "[System.Environment]::GetEnvironmentVariable(", .pattern = .quoted },
};

const swift = [_]Accessor{
    .{ .prefix = "ProcessInfo.processInfo.environment[", .pattern = .quoted },
    .{ .prefix = "getenv(", .pattern = .quoted },
};

const objc = [_]Accessor{
    .{ .prefix = "getenv(", .pattern = .quoted },
    .{ .prefix = "environment] objectForKey:@", .pattern = .quoted },
};

const dart = [_]Accessor{
    .{ .prefix = "Platform.environment[", .pattern = .quoted },
};

const elixir = [_]Accessor{
    .{ .prefix = "System.get_env(", .pattern = .quoted },
    .{ .prefix = "System.fetch_env!(", .pattern = .quoted },
    .{ .prefix = "System.fetch_env(", .pattern = .quoted },
};

const erlang = [_]Accessor{
    .{ .prefix = "os:getenv(", .pattern = .quoted },
};

const haskell = [_]Accessor{
    .{ .prefix = "getEnv ", .pattern = .quoted },
    .{ .prefix = "lookupEnv ", .pattern = .quoted },
};

const ocaml = [_]Accessor{
    .{ .prefix = "Sys.getenv ", .pattern = .quoted },
    .{ .prefix = "Sys.getenv_opt ", .pattern = .quoted },
    .{ .prefix = "Unix.getenv ", .pattern = .quoted },
};

const lua = [_]Accessor{
    .{ .prefix = "os.getenv(", .pattern = .quoted },
};

const r = [_]Accessor{
    .{ .prefix = "Sys.getenv(", .pattern = .quoted },
};

const julia = [_]Accessor{
    .{ .prefix = "ENV[", .pattern = .quoted },
    .{ .prefix = "get(ENV, ", .pattern = .quoted },
};

const crystal = [_]Accessor{
    .{ .prefix = "ENV[", .pattern = .quoted },
    .{ .prefix = "ENV.fetch(", .pattern = .quoted },
};

const nim = [_]Accessor{
    .{ .prefix = "getEnv(", .pattern = .quoted },
    .{ .prefix = "existsEnv(", .pattern = .quoted },
};

const dlang = [_]Accessor{
    .{ .prefix = "environment[", .pattern = .quoted },
    .{ .prefix = "environment.get(", .pattern = .quoted },
    // NOTE: `"FOO" in environment` puts the key BEFORE the prefix; a forward
    // prefix matcher cannot catch it. Accept the miss or special-case it.
};

const vlang = [_]Accessor{
    .{ .prefix = "os.getenv(", .pattern = .quoted },
};

const fortran = [_]Accessor{
    .{ .prefix = "get_environment_variable(", .pattern = .quoted },
};

const pascal = [_]Accessor{
    .{ .prefix = "GetEnvironmentVariable(", .pattern = .quoted },
    .{ .prefix = "GetEnv(", .pattern = .quoted },
};

const tcl = [_]Accessor{
    .{ .prefix = "$env(", .pattern = .parened },
};

const nushell = [_]Accessor{
    .{ .prefix = "$env.", .pattern = .ident },
};

// const shell = [_]Accessor{
//     .{ .prefix = "${", .pattern = .braced },
//     .{ .prefix = "$", .pattern = .ident },
// };

pub const extensions = std.StaticStringMap([]const Accessor).initComptime(.{
    .{ "js", &javascript },   .{ "mjs", &javascript }, .{ "cjs", &javascript },
    .{ "jsx", &javascript },  .{ "ts", &javascript },  .{ "tsx", &javascript },
    .{ "mts", &javascript },  .{ "cts", &javascript }, .{ "py", &python },
    .{ "pyi", &python },      .{ "pyw", &python },     .{ "go", &go },
    .{ "rs", &rust },         .{ "rb", &ruby },        .{ "rake", &ruby },
    .{ "gemspec", &ruby },    .{ "zig", &zig },        .{ "java", &java },
    .{ "kt", &java },         .{ "kts", &java },       .{ "scala", &scala },
    .{ "sc", &scala },        .{ "groovy", &java },    .{ "gradle", &java },
    .{ "clj", &clojure },     .{ "cljs", &clojure },   .{ "cljc", &clojure },
    .{ "c", &c },             .{ "cpp", &cpp },        .{ "cc", &cpp },
    .{ "cxx", &cpp },         .{ "hpp", &cpp },        .{ "hh", &cpp },
    .{ "hxx", &cpp },         .{ "h", &cpp },          .{ "m", &objc },
    .{ "mm", &objc },         .{ "cs", &dotnet },      .{ "fs", &dotnet },
    .{ "fsx", &dotnet },      .{ "fsi", &dotnet },     .{ "vb", &visualbasic },
    .{ "php", &php },         .{ "pl", &perl },        .{ "pm", &perl },
    .{ "t", &perl },          .{ "ps1", &powershell }, .{ "psm1", &powershell },
    .{ "psd1", &powershell }, .{ "swift", &swift },    .{ "dart", &dart },
    .{ "ex", &elixir },       .{ "exs", &elixir },     .{ "erl", &erlang },
    .{ "hrl", &erlang },      .{ "hs", &haskell },     .{ "lhs", &haskell },
    .{ "ml", &ocaml },        .{ "mli", &ocaml },      .{ "lua", &lua },
    .{ "r", &r },             .{ "jl", &julia },       .{ "cr", &crystal },
    .{ "nim", &nim },         .{ "d", &dlang },        .{ "v", &vlang },
    .{ "f90", &fortran },     .{ "f95", &fortran },    .{ "f03", &fortran },
    .{ "f08", &fortran },     .{ "for", &fortran },    .{ "f", &fortran },
    .{ "pas", &pascal },      .{ "pp", &pascal },      .{ "dpr", &pascal },
    .{ "tcl", &tcl },         .{ "nu", &nushell },
    // .{ "sh", &shell }, .{ "bash", &shell }, .{ "zsh", &shell }, .{ "ksh", &shell },
});

File: src/accessors.c
#include "accessors.h"
#include "macros.h"
#include <string.h>

static const accessor_t javascript[] = {
    ACCESSOR("process.env.", ident),      ACCESSOR("process.env[", quoted),  ACCESSOR("import.meta.env.", ident),
    ACCESSOR("import.meta.env[", quoted), ACCESSOR("Deno.env.get(", quoted), ACCESSOR("Bun.env.", ident),
    ACCESSOR("Bun.env[", quoted),
};

static const accessor_t python[] = {
    ACCESSOR("os.environ.setdefault(", quoted),
    ACCESSOR("getenv(", quoted),
    ACCESSOR("environ[", quoted),
    ACCESSOR("environ.get(", quoted),
};

static const accessor_t go[] = {
    ACCESSOR("os.Getenv(", quoted),
    ACCESSOR("os.LookupEnv(", quoted),
};

static const accessor_t rust[] = {
    ACCESSOR("env::var(", quoted),
    ACCESSOR("env::var_os(", quoted),
    ACCESSOR("env!(", quoted),
    ACCESSOR("option_env!(", quoted),
};

static const accessor_t ruby[] = {
    ACCESSOR("ENV[", quoted),
    ACCESSOR("ENV.fetch(", quoted),
};

static const accessor_t zig[] = {
    ACCESSOR("getenv(", quoted),         ACCESSOR("getEnvVarOwned(", quoted),
    ACCESSOR("hasEnvVar(", quoted),      ACCESSOR("hasEnvVarConstant(", quoted),
    ACCESSOR("parseEnvVarInt(", quoted),
};

static const accessor_t java[] = {
    ACCESSOR("System.getenv(", quoted),
};

static const accessor_t scala[] = {
    ACCESSOR("sys.env(", quoted),
    ACCESSOR("sys.env.get(", quoted),
    ACCESSOR("System.getenv(", quoted),
};

static const accessor_t clojure[] = {
    ACCESSOR("System/getenv ", quoted),
    ACCESSOR("(getenv ", quoted),
};

static const accessor_t c[] = {
    ACCESSOR("getenv(", quoted),
    ACCESSOR("secure_getenv(", quoted),
};

static const accessor_t cpp[] = {
    ACCESSOR("getenv(", quoted),
};

static const accessor_t dotnet[] = {
    ACCESSOR("Environment.GetEnvironmentVariable(", quoted),
};

static const accessor_t visualbasic[] = {
    ACCESSOR("Environment.GetEnvironmentVariable(", quoted),
    ACCESSOR("Environ(", quoted),
};

static const accessor_t php[] = {
    ACCESSOR("getenv(", quoted),
    ACCESSOR("$_ENV[", quoted),
    ACCESSOR("$_SERVER[", quoted),
};

static const accessor_t perl[] = {
    ACCESSOR("$ENV{", braced),
};

static const accessor_t powershell[] = {
    ACCESSOR("$env:", ident),
    ACCESSOR("[Environment]::GetEnvironmentVariable(", quoted),
    ACCESSOR("[System.Environment]::GetEnvironmentVariable(", quoted),
};

static const accessor_t swift[] = {
    ACCESSOR("ProcessInfo.processInfo.environment[", quoted),
    ACCESSOR("getenv(", quoted),
};

static const accessor_t objc[] = {
    ACCESSOR("getenv(", quoted),
    ACCESSOR("environment] objectForKey:@", quoted),
};

static const accessor_t dart[] = {
    ACCESSOR("Platform.environment[", quoted),
};

static const accessor_t elixir[] = {
    ACCESSOR("System.get_env(", quoted),
    ACCESSOR("System.fetch_env!(", quoted),
    ACCESSOR("System.fetch_env(", quoted),
};

static const accessor_t erlang[] = {
    ACCESSOR("os:getenv(", quoted),
};

static const accessor_t haskell[] = {
    ACCESSOR("getEnv ", quoted),
    ACCESSOR("lookupEnv ", quoted),
};

static const accessor_t ocaml[] = {
    ACCESSOR("Sys.getenv ", quoted),
    ACCESSOR("Sys.getenv_opt ", quoted),
    ACCESSOR("Unix.getenv ", quoted),
};

static const accessor_t lua[] = {
    ACCESSOR("os.getenv(", quoted),
};

static const accessor_t r[] = {
    ACCESSOR("Sys.getenv(", quoted),
};

static const accessor_t julia[] = {
    ACCESSOR("ENV[", quoted),
    ACCESSOR("get(ENV, ", quoted),
};

static const accessor_t crystal[] = {
    ACCESSOR("ENV[", quoted),
    ACCESSOR("ENV.fetch(", quoted),
};

static const accessor_t nim[] = {
    ACCESSOR("getEnv(", quoted),
    ACCESSOR("existsEnv(", quoted),
};

static const accessor_t dlang[] = {
    ACCESSOR("environment[", quoted),
    ACCESSOR("environment.get(", quoted),
};

static const accessor_t vlang[] = {
    ACCESSOR("os.getenv(", quoted),
};

static const accessor_t fortran[] = {
    ACCESSOR("get_environment_variable(", quoted),
};

static const accessor_t pascal[] = {
    ACCESSOR("GetEnvironmentVariable(", quoted),
    ACCESSOR("GetEnv(", quoted),
};

static const accessor_t tcl[] = {
    ACCESSOR("$env(", parened),
};

static const accessor_t nushell[] = {
    ACCESSOR("$env.", ident),
};

static const accessor_t yaml[] = {
    ACCESSOR("${", expansion),
};

static const ext_entry extensions[] = {
    ENTRY("js", javascript),
    ENTRY("mjs", javascript),
    ENTRY("cjs", javascript),
    ENTRY("jsx", javascript),
    ENTRY("ts", javascript),
    ENTRY("tsx", javascript),
    ENTRY("mts", javascript),
    ENTRY("cts", javascript),
    ENTRY("py", python),
    ENTRY("pyi", python),
    ENTRY("pyw", python),
    ENTRY("go", go),
    ENTRY("rs", rust),
    ENTRY("rb", ruby),
    ENTRY("rake", ruby),
    ENTRY("gemspec", ruby),
    ENTRY("zig", zig),
    ENTRY("java", java),
    ENTRY("kt", java),
    ENTRY("kts", java),
    ENTRY("groovy", java),
    ENTRY("gradle", java),
    ENTRY("scala", scala),
    ENTRY("sc", scala),
    ENTRY("clj", clojure),
    ENTRY("cljs", clojure),
    ENTRY("cljc", clojure),
    ENTRY("c", c),
    ENTRY("cpp", cpp),
    ENTRY("cc", cpp),
    ENTRY("cxx", cpp),
    ENTRY("hpp", cpp),
    ENTRY("hh", cpp),
    ENTRY("hxx", cpp),
    ENTRY("h", cpp),
    ENTRY("m", objc),
    ENTRY("mm", objc),
    ENTRY("cs", dotnet),
    ENTRY("fs", dotnet),
    ENTRY("fsx", dotnet),
    ENTRY("fsi", dotnet),
    ENTRY("vb", visualbasic),
    ENTRY("php", php),
    ENTRY("pl", perl),
    ENTRY("pm", perl),
    ENTRY("t", perl),
    ENTRY("ps1", powershell),
    ENTRY("psm1", powershell),
    ENTRY("psd1", powershell),
    ENTRY("swift", swift),
    ENTRY("dart", dart),
    ENTRY("ex", elixir),
    ENTRY("exs", elixir),
    ENTRY("erl", erlang),
    ENTRY("hrl", erlang),
    ENTRY("hs", haskell),
    ENTRY("lhs", haskell),
    ENTRY("ml", ocaml),
    ENTRY("mli", ocaml),
    ENTRY("lua", lua),
    ENTRY("r", r),
    ENTRY("jl", julia),
    ENTRY("cr", crystal),
    ENTRY("nim", nim),
    ENTRY("d", dlang),
    ENTRY("v", vlang),
    ENTRY("f90", fortran),
    ENTRY("f95", fortran),
    ENTRY("f03", fortran),
    ENTRY("f08", fortran),
    ENTRY("for", fortran),
    ENTRY("f", fortran),
    ENTRY("pas", pascal),
    ENTRY("pp", pascal),
    ENTRY("dpr", pascal),
    ENTRY("tcl", tcl),
    ENTRY("nu", nushell),
    ENTRY("yaml", yaml),
    ENTRY("yml", yaml),
};

static const size_t ext_len = ARR_LEN(extensions);

const ext_entry *get_scan_extension(const char *ext) {
    for (size_t i = 0; i < ext_len; ++i) {
        if (strcmp(ext, extensions[i].ext) == 0) {
            return &extensions[i];
        }
    }
    return NULL;
}

const file_ext_t *get_file_extension(const file_ext_map_t *map, const char *ext) {
    for (size_t i = 0; i < map->count; ++i) {
        if (strcmp(map->items[i].ext, ext) == 0) {
            return &map->items[i];
        }
    }
    return NULL;
}

void append_file_extension(file_ext_map_t *map, const ext_entry *entry) {
    if (get_file_extension(map, entry->ext) != NULL) {
        return;
    }

    file_ext_t match = {
        .ext = entry->ext,
        .accessors = entry->accessors,
        .accessor_count = entry->count,
    };

    DYN_ARR_APPEND(map, match);
}
